Source/Core/Core/State.cpp

```diff
@@ -412,111 +425,82 @@ static void CompressAndDumpState(Core::System& system, CompressAndDumpState_args
     }
-    else
+    else if (File::Exists(dtmname))
     {
-      const std::filesystem::path temp_path(filename);
-      Core::DisplayMessage(fmt::format("Saved State to {}", temp_path.filename().string()), 2000);
+      if (!File::Rename(dtmname, last_state_dtmname))
+        Core::DisplayMessage("Failed to move previous state's dtm to state undo backup", 1000);
     }
   }
-}
-
-void SaveAs(Core::System& system, const std::string& filename, bool wait)
-{
-  std::unique_lock lk(s_load_or_save_in_progress_mutex, std::try_to_lock);
-  if (!lk)
-    return;
-
-  Core::RunOnCPUThread(
-      system,
-      [&] {
-        {
-          std::lock_guard lk_(s_state_writes_in_queue_mutex);
-          ++s_state_writes_in_queue;
-        }
 
-        // Measure the size of the buffer.
-        u8* ptr = nullptr;
-        PointerWrap p_measure(&ptr, 0, PointerWrap::Mode::Measure);
-        DoState(system, p_measure);
-        const size_t buffer_size = ptr - (u8*)(nullptr);
+  auto& movie = system.GetMovie();
+  if ((movie.IsMovieActive()) && !movie.IsJustStartingRecordingInputFromSaveState())
+    movie.SaveRecording(dtmname);
+  else if (!movie.IsMovieActive())
+    File::Delete(dtmname);
 
-        // Then actually do the write.
-        Common::UniqueBuffer<u8> current_buffer(buffer_size);
-        ptr = current_buffer.data();
-        PointerWrap p(&ptr, buffer_size, PointerWrap::Mode::Write);
-        DoState(system, p);
+  // Move written state to final location.
+  // TODO: This should also be atomic. This is possible on all systems, but needs a special
+  // implementation of IOFile on Windows.
+  if (!f.Close())
+    Core::DisplayMessage("Failed to close state file", 2000);
 
-        if (p.IsWriteMode())
-        {
-          Core::DisplayMessage("Saving State...", 1000);
+  if (!File::Rename(temp_filename, filename))
+  {
+    Core::DisplayMessage("Failed to rename state file", 2000);
+  }
+  else
+  {
+    const std::filesystem::path temp_path(filename);
+    Core::DisplayMessage(fmt::format("Saved State to {}", temp_path.filename().string()), 2000);
+  }
+}
 
-          std::shared_ptr<Common::Event> sync_event;
+static void SaveAsFromCore(Core::System& system, std::string filename)
+{
+  // Try with a buffer a bit larger than the previous state.
+  // This will often avoid the "Measure" step.
+  const auto buffer_size_estimate = std::size_t(s_last_state_size) * 110 / 100;
+  Common::UniqueBuffer<u8> buffer{buffer_size_estimate};
 
-          CompressAndDumpState_args save_args;
-          save_args.buffer = std::move(current_buffer);
-          save_args.filename = filename;
-          if (wait)
-          {
-            sync_event = std::make_shared<Common::Event>();
-            save_args.state_write_done_event = sync_event;
-          }
+  if (const auto actual_size = SaveToBuffer(system, buffer))
+  {
+    // Adjust the oversized buffer down to the actual size.
+    buffer.assign(buffer.extract().first, actual_size);
```

Comment:
You deleted `resize` in `BufferBase` to avoid unnecessary copies, but would it be worth adding `ShrinkToSize(size_t)` that could only decrease the size (or keep it the same)?